File: src/main/java/org/arif/DAY_01/Day01_TwoSum.java
package org.arif.DAY_01;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class Day01_TwoSum {
    public static void main(String[] args) {
        int[] nums = {2, 7, 11, 5};
        int target = 9;
        int[] result = twoSum(nums, target);
        System.out.println(Arrays.toString(result));
    }

    private static int[] twoSum(int[] nums, int target) {
        for (int i = 0; i < nums.length - 1; i++) {
            for (int j = i+1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    return new int[]{i, j};
                }
            }
        }
        throw new IllegalArgumentException("INVALID_INPUT");
    }
    private static int[] twoSum02(int[] nums, int target) {

        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int remain = target - nums[i];
            if (map.containsKey(remain)) {
                return new int[]{map.get(remain), i};
            }
            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("INVALID_INPUT");
    }
    private static int[] twoSum03(int[] nums, int target) {

        for (int i = 0; i < nums.length; i++) {

        }
        throw new IllegalArgumentException("INVALID_INPUT");
    }

}
/**
 * Given an array of integers nums and an integer target,
 * return indices of the two numbers such that they add up to target.
 * You may assume that each input would have exactly one solution,
 * and you may not use the same element twice.
 * You can return the answer in any order.
 * <p>
 *
 * Example 1:
 * Input: nums = [2,7,11,15], target = 9
 * Output: [0,1]
 * Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
 * <p>
 *
 * Example 2:
 *
 * Input: nums = [3,2,4], target = 6
 * Output: [1,2]
 * <p>
 *
 * Example 3:
 *
 * Input: nums = [3,3], target = 6
 * Output: [0,1]
 */

File: src/main/java/org/arif/DAY_01/Day01_TwoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
/*const twoSum = function (nums, target) {
    let result = [];
    for (let i = 0; i < nums.length - 1; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) {
                result.push(i);
                result.push(j);
            }
        }
    }
    return result;
};*/

const nums = [2, 7, 11, 5];
const target = 9;

//const result = twoSum(nums, target);
//console.log(result);

// another approach
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSum2 = (nums, target) => {
    const result = []
    const n = nums.length
    const map = new Map();

    for (let i = 0; i < n; i++) {
        let remained = target - nums[i];

        if (map.has(remained)) {
            result.push(map.get(remained));
            result.push(i);
            break;
        }
        map.set(nums[i], i);
    }

            return result;
};

const result2 = twoSum2(nums, target)
console.log(result2)